Add torch._library.register_fake_class to fakify torchBind class

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122622
* #122621
* #122620
* #122619

This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.

ydwu4 committed:
Add torch._library.impl_abstract_class to fakify torchBind class

[ghstack-poisoned]

zou3519 commented on test/export/test_torchbind.py:
```diff
@@ -16,6 +16,47 @@
+
+            @classmethod
+            def from_concrete(cls, real_tq):
+                return cls(real_tq.clone_queue())
```
I expected this to create a FakeTensorQueue with FakeTensors. Instead, it looks like we're cloning real Tensors and putting them into the FakeTensorQueue. What's up?

ydwu4 replied:
Yeah, you're right. This example didn't fakify the internal states of FakeTensorQueue, and we should do it in this PR. I can think of two possible approaches of doing it: 

1. **We do the fakification.** Specifically, users deal with real tensors and use whatever ways they have/need to retrieve the state from the real tensorQueue and create a FakeTensorQueue instance.  We fakify the FakeTensorQueue further by recursively applying something like the wrap_fake of make_fx's to the FakeTensorQueue's attributes.

2. **Users do the fakification.** Users need to use the functions that we have in wrap_fake (or we could provide some user-facing fakification API) to create the FakeTensorQueue and we'll use the FakeTensoQueue as it is.

I'm more in favor of 1. because users may not always write the correct fakification logic. For example, forget to use the fakification API at all, doesn't fakify some tensor/integer, or doesn't recursively fakify the ScriptObject nested in current one. One example is an `add_tensor` method that returns `self.real + x`, where `x` is fake but `self.real` is not properly fakified).

ydwu4 replied:
Resolved online. decided to use Option 2 as it's more explicit and has a clear contract.

zou3519 commented on test/export/test_torchbind.py:
```diff
@@ -16,6 +16,47 @@
+
+            def size(self):
+                return len(self.queue)
+
```
Meta comment: It's kind of difficult to review this PR by itself, the PRs stacked on top of it matter

ydwu4 committed:
Update on "Add torch._library.impl_abstract_class to fakify torchBind class"


This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.



[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch._library.impl_abstract_class to fakify torchBind class"


This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.



[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch._library.impl_abstract_class to fakify torchBind class"


This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.



[ghstack-poisoned]

zou3519 commented on test/export/test_torchbind.py:
```diff
@@ -16,6 +19,66 @@
 @skipIfTorchDynamo("torchbind not supported with dynamo yet")
 class TestExportTorchbind(TestCase):
+    def setUp(self):
+        @torch._library.impl_abstract_class("_TorchScriptTesting::_Foo")
```
Naming: probably just call it "register_fake_class". I'm planning on changing `impl_abstract` to "register_fake"

zou3519 commented on torch/_library/abstract_impl.py:
```diff
@@ -204,3 +205,38 @@ def new_dynamic_size(self, *, min=0, max=None) -> torch.SymInt:
         )
         return result
+
+    def create_fake_tensor(self, tensor: torch.Tensor):
```
nit: maybe `to_fake_tensor`? Sometimes we're not "creating" a new Tensor (if a tensor is a view on another one, and we have already fakeified the other one, then this will be a view on the fakeified tensor)

zou3519 commented on torch/_library/abstract_impl.py:
```diff
@@ -204,3 +205,38 @@ def new_dynamic_size(self, *, min=0, max=None) -> torch.SymInt:
+            >>>     @classmethod
+            >>>     def from_real(cls, real_tq):
+            >>>         ctx = torch.library.get_ctx()
+            >>>         fake_queue = [ctx.create_fake_tensor(t) for t in real_tq.clone_queue()]
```
We don't actually need to `clone_queue` anymore, right? we just need to call `create_fake_tensor` for each tensor in real_tq.

ydwu4 replied:
the list attribute is not exposed to python, we need some ways of getting the queue out. Maybe we could just change to get_queue() to make this intention more explicity so that users won't have the wrong impression that the states have to be copied.

zou3519 commented on torch/_library/abstract_impl_class.py:
```diff
@@ -0,0 +1,229 @@
+log = logging.getLogger(__name__)
+
+
+class AbstractClassRegistry:
```
Let's do "abstract" -> "fake" (sorry for the naming churn)

zou3519 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1153,6 +1159,8 @@ def wrap_fake(x):
             # NB: don't match on bools
             elif type(x) is int and tracing_mode == "symbolic":
                 return shape_env.create_symintnode(shape_env.create_symbol(x, source, positive=None), hint=x, source=source)
+            elif isinstance(x, torch.ScriptObject):
```
as discussed offline, we probably want to assert if we see a FakeScriptObject

ydwu4 committed:
Update on "Add torch._library.impl_abstract_class to fakify torchBind class"


This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.



[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch._library.impl_abstract_class to fakify torchBind class"


This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.



[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch._library.impl_abstract_class to fakify torchBind class"


This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.



[ghstack-poisoned]

zou3519 commented on test/export/test_torchbind.py:
```diff
@@ -351,6 +414,188 @@ def forward(self, arg0_1, attr, arg1_1):
+            ).fill_(-1)
+        )
+        x = torch.ones(2, 3)
+        with torch._higher_order_ops.torchbind.enable_torchbind_tracing():
```
Sorry, I forgot why -- why do we need to enable this?

ydwu4 replied:
It's only useful for tracing with real script object. In our case, we've fakified the script object. I should remove it.

zou3519 commented on test/export/test_torchbind.py:
```diff
@@ -351,6 +414,188 @@ def forward(self, arg0_1, attr, arg1_1):
+
+
+@skipIfTorchDynamo("torchbind not supported with dynamo yet")
+class TestImplAbstractClass(TestCase):
```
nit: TestRegisterFakeClass ?

ydwu4 replied:
oops

zou3519 commented on test/export/test_torchbind.py:
```diff
@@ -351,6 +414,188 @@ def forward(self, arg0_1, attr, arg1_1):
+                x, y = foo_obj.__getstate__()
+                return cls(x, y)
+
+        with self.assertRaisesRegex(RuntimeError, "already registered"):
+            torch._library.register_fake_class("_TorchScriptTesting::_Foo", FakeFoo)
```
(no action required) If the user registers two, it's better to just accept the second one. I'm going to update `impl_abstract` to do that at some point. If it's not too difficult to make that happen in register_abstract_class now, we should.

zou3519 replied:
(registering two Fake classes should result in a warning but allow it through; this is useful for situations like colab)

zou3519 commented on torch/_library/fake_class_registry.py:
```diff
@@ -0,0 +1,227 @@
+    def __init__(self):
+        self._registered_class: Dict[str, Any] = {}
+
+    def has_impl(self, full_qualname: str):
```
can you add return type hints for these?

zou3519 commented on torch/_library/fake_class_registry.py:
```diff
@@ -0,0 +1,227 @@
+        self._check_registered(full_qualname)
+        return self._registered_class.pop(full_qualname)
+
+    def clear(self):
+        self._registered_class.clear()
+
+    def _check_registered(self, full_qualname: str):
```
ditto on type hints

zou3519 commented on torch/_library/fake_class_registry.py:
```diff
@@ -0,0 +1,227 @@
+global_fake_class_registry = FakeClassRegistry()
+
+
+def to_fake_obj(fake_mode, x: torch.ScriptObject):
```
return type hint

zou3519 commented on torch/_library/fake_class_registry.py:
```diff
@@ -0,0 +1,227 @@
+
+    """
+
+    def inner(fake_class):
```
We should do some protocol type hinting here. e.g.:
```
def inner(fake_class: FakeClassProto) -> FakeClassProto
```
FakeClassProto is some protocol that requires a staticmethod from_real method

zou3519 replied:
https://peps.python.org/pep-0544/

ydwu4 replied:
sounds good! I'll check it out and try to apply it.

zou3519 replied:
oh and you can apply this to the input fake_class too if it is not None

zou3519 commented on torch/_library/abstract_impl.py:
```diff
@@ -204,3 +205,39 @@ def new_dynamic_size(self, *, min=0, max=None) -> torch.SymInt:
+        """
+        Creates a fake tensor from a concrete tensor.
+
+        This is useful for writing the fake implementation (which is necessary
+        for torch.compile) for custom class. Users need to implement a from_real method that takes
+        a real custom object and creates a fake custom object. Users can use this API to create
+        fake tensors for the tensor states in the custom object.
```
Can you add a line saying that this isn't needed for `impl_abstract`? But may be needed for `register_fake_class`?

zou3519 commented on torch/_higher_order_ops/torchbind.py:
```diff
@@ -69,6 +78,21 @@ def inner(mode, *args, **kwargs):
+            ns, class_name = _ns_and_class_name(
+                obj._type().qualified_name()  # type: ignore[attr-defined]
+            )
+            log.warning(
+                "Tracing torchbind method %s.%s with real ScriptObject. This may"
+                " cause the original object being mutated. If this is not intended,"
+                ' You can register a fake class with torch._library.register_fake_class("%s::%s").',
```
mm, we should probably error out on this if FakeTensorMode is active

ydwu4 replied:
We cannot do it in this diff yet because many tests in current export is still tracing with real ScriptObject. We can do this when we turn on fakification by default.

zou3519 replied:
ya sounds good

zou3519 reviewed (approved):
LGTM. Just some minor comments around safety and APIs. Also, type hints on everything will help future readers

ydwu4 committed:
Update on "Add torch._library.impl_abstract_class to fakify torchBind class"


This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.



[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch._library.register_fake_class to fakify torchBind class"


This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.



[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch._library.register_fake_class to fakify torchBind class"


This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.



[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "Add torch._library.register_fake_class to fakify torchBind class"


This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.



[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch._library.register_fake_class to fakify torchBind class"


This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.



[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch._library.register_fake_class to fakify torchBind class"


This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.



[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch._library.register_fake_class to fakify torchBind class"


This PR only adds abstract class registration logic without touching existing tests so they still trace with real script object. The added tests are only for registration APIs and test error messages.

Our design is that the abstract implementation should be in Python. This is much better in terms of usability. But this also has implications for custom op that takes script object as input, which is detailed later in this stack.



[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
